## Task 1: Extract data from Wikipedia and create a Dataframe

#### 1. Import libraries

In [1]:
import requests
import bs4
import pandas as pd
from bs4 import BeautifulSoup

#### 2. Create dataframe

In [2]:
# Get html raw data from Wikipedia page
r = requests.get('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M')

In [3]:
# Read html as text
html = BeautifulSoup(r.text)

In [4]:
# Create dataframe by looping over non-empty and Tag-type rows
postcode = []
borough = []
neighborhood = []

for tr in html.table.tbody:
    values = []    
    for td in tr:
        if not isinstance(td, bs4.element.Tag):
            continue
        if td.name == 'td':
            values.append(td.text.strip())
    if len(values) == 0:
        continue
    postcode.append(values[0])
    borough.append(values[1])
    neighborhood.append(values[2])
    
df = pd.DataFrame({
    "PostalCode": postcode,
    "Borough": borough,
    "Neighborhood": neighborhood
})

In [5]:
# Unclean dataframe
df

PostalCode           Borough  \
0          M1A      Not assigned   
1          M2A      Not assigned   
2          M3A        North York   
3          M4A        North York   
4          M5A  Downtown Toronto   
..         ...               ...   
175        M5Z      Not assigned   
176        M6Z      Not assigned   
177        M7Z      Not assigned   
178        M8Z         Etobicoke   
179        M9Z      Not assigned   

                                          Neighborhood  
0                                         Not assigned  
1                                         Not assigned  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
..                                                 ...  
175                                       Not assigned  
176                                       Not assigned  
177                                       Not assigned  
178  Mimico NW, The Queensway West, South of Bloor,...  
179                                       Not assigned  

[180 rows x 3 columns]

#### 3. Modify the dataframe as instructed

#### 3.1. Remove cells with a borough that is **Not assigned**

In [12]:
df = df[df['Borough'] != 'Not assigned']

In [13]:
df

PostalCode           Borough  \
2          M3A        North York   
3          M4A        North York   
4          M5A  Downtown Toronto   
5          M6A        North York   
6          M7A  Downtown Toronto   
..         ...               ...   
160        M8X         Etobicoke   
165        M4Y  Downtown Toronto   
168        M7Y      East Toronto   
169        M8Y         Etobicoke   
178        M8Z         Etobicoke   

                                          Neighborhood  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
5                     Lawrence Manor, Lawrence Heights  
6          Queen's Park, Ontario Provincial Government  
..                                                 ...  
160      The Kingsway, Montgomery Road, Old Mill North  
165                               Church and Wellesley  
168  Business reply mail Processing Centre, South C...  
169  Old Mill South, King's Mill Park, Sunnylea, Hu...  
178  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

#### 3.2. Check if there are any **Not assigned** neighborhoods

In [19]:
df[df['Neighborhood'] == 'Not assigned']

Empty DataFrame
Columns: [PostalCode, Borough, Neighborhood]
Index: []

#### 3.3. Check the size of the final dataframe

In [20]:
df.shape

(103, 3)

##### The final dataframe has 103 rows.

#### **Optional:** Export the final dataframe so that it can be used for Task 2 and 3.

In [21]:
df.to_csv('can_df.csv', index=False)